harness/tests/experiment/pytorch/test_pytorch_trial.py

```diff
@@ -601,19 +605,36 @@ def make_workloads() -> workload.Stream:
         )
         controller.run()
 
-    @pytest.mark.skipif(not HAVE_APEX or not torch.cuda.is_available(), reason="no gpu available")
+    @pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
     @pytest.mark.gpu
-    def test_apex_amp(self) -> None:
-        apex.amp.register_float_function(torch, "sigmoid")
+    @pytest.mark.parametrize(
+        "trial_class",
+        [
+            pytorch_onevar_model.OneVarApexAMPTrial,
+            pytorch_onevar_model.OneVarAutoAMPTrial,
+            pytorch_onevar_model.OneVarManualAMPTrial,
+        ],
+        ids=[
+            "apex",
+            "autocast",
+            "manual",
+        ],
+    )
+    def test_amp(self, trial_class) -> None:
+        if trial_class is pytorch_onevar_model.OneVarApexAMPTrial and not HAVE_APEX:
+            pytest.skip("Apex not available")
 
-        def make_workloads() -> workload.Stream:
-            trainer = utils.TrainAndValidate()
-            yield from trainer.send(steps=1, validation_freq=1, scheduling_unit=1)
+        if trial_class is pytorch_onevar_model.OneVarApexAMPTrial:
```

Comment:
You can parameterize multiple values at once to avoid this awkward else/if'ing:

```
@parameterize(
    "class,output_float16,scale_changed",
    [(OneVarTrialWithApexAmp,False,True), ...],
)
```

Reply:
We can definitely do this (and I have the feeling we might as well), but the reason I made the `assert_*` flags be directly derived from the trial class is because - when either/both are False - it's a choice made out of necessary, not, say, preference for typical use-cases. Ideally we'd have checks for casting and changes of scale for all three trial classes.

That might not be clear, but I think this is an ok change anyway. A plus is that what exactly is being checked in the three test variations is made clear when running them (without looking at the code).

Reply:
Actually, the plus I mentioned isn't quite true since we have the "apex", "autocast", and "manual" ids.